src/data/simple_dmatrix.h

```diff
@@ -47,20 +47,28 @@ class SimpleDMatrix : public DMatrix {
     auto& data_vec = mat.page_.data.HostVector();
     uint64_t inferred_num_columns = 0;
 
+    size_t num_lines_processed = 0;
+
+    using OffsetVecT = std::remove_reference<decltype(offset_vec)>::type::value_type;
+    std::vector<std::vector<OffsetVecT>> tmp_thread_ptr;
+
     adapter->BeforeFirst();
     // Iterate over batches of input data
     while (adapter->Next()) {
       auto &batch = adapter->Value();
-      common::ParallelGroupBuilder<
-        Entry, std::remove_reference<decltype(offset_vec)>::type::value_type>
-        builder(&offset_vec, &data_vec);
-      builder.InitBudget(0, nthread);
+
+      // Batch data and its offsets
+      std::vector<OffsetVecT> batch_offset_vec;
+      std::vector<Entry> batch_data_vec;
+      common::ParallelGroupBuilder<Entry, OffsetVecT> builder(&batch_offset_vec,
+                                                              &batch_data_vec,
+                                                              &tmp_thread_ptr);
 
       // First-pass over the batch counting valid elements
       size_t num_lines = batch.Size();
+      builder.InitBudget(num_lines, nthread);
```

Comment:
`num_lines` != number of rows. I built this as an abstraction that can deal with both CSC/CSR sparse matrices. In the case of CSC num_lines will be the number of columns.

Reply:
Perhaps instead of this we can use heuristic to initialise the expected number of rows (that should be exact for all current cases). Take the last element of the last "line" and use its row index.

Reply:
@RAMitchell Should we revert the commit and prepare a new one?  Seems to be cleaner way.

Reply:
the csX adapters that are present today seem to be single batch adapters, in that they return only one batch in total. hence, the offsets to the row indices by the number of rows processed thus far may at best be moot. this *perhaps* could be the reason why all these tests are passing.

that said, it doesn't preclude someone to create an adapter that can return a csc formatted data as multiple batches in the future, and we should handle this, which the original version does not.

please check the new version, as this is being handled slightly differently. i have also added a csc adapter test that forces the adapters to infer sizes, thus in a way also ascertaining the integrity of the parallel build.

Reply:
@trivialfis reverting is unnecessary.